common/core/desktop/src/kmx/kmx_xstring.cpp

```diff
@@ -167,49 +169,28 @@ PKMX_WCHAR km::kbp::kmx::decxstr(PKMX_WCHAR p, PKMX_WCHAR pStart)
     return p;
   }
 
-  if(p == pStart) return p; // Don't allow test before pStart
+  if (p == pStart) {
+    // Don't allow test before pStart
+    return p;
+  }
 
-  if(*p >= 0xDC00 && *p <= 0xDFFF && *(p-1) >= 0xD800 && *(p-1) <= 0xDBFF)
-  {
+  if(*p >= 0xDC00 && *p <= 0xDFFF && *(p-1) >= 0xD800 && *(p-1) <= 0xDBFF) {
     return p-1;
   }
-  else if(*(p-1) == UC_SENTINEL) return p-1;
-  else if(p > pStart+1 && *(p-2) == UC_SENTINEL)
-  {
-    switch(*(p-1))
-    {
-      case CODE_ANY:
-      case CODE_NOTANY:
-      case CODE_USE:
-      case CODE_DEADKEY:
-      case CODE_CLEARCONTEXT:
-      case CODE_CALL:
-      case CODE_CONTEXTEX:
-      case CODE_RESETOPT:
-      case CODE_SAVEOPT:
-        return p-2;
-    }
-  }
-  else if(p > pStart+2 && *(p-3) == UC_SENTINEL)
-  {
-    switch(*(p-2))
-    {
-      case CODE_INDEX:
-      case CODE_SETOPT:
-      case CODE_SETSYSTEMSTORE:
-        return p-3;
-    }
+
+  if (*(p - 1) == UC_SENTINEL) {
+    // TODO: validate that *p <= CODE_LASTCODE? What do we do if it isn't?
+    return p - 1;
   }
-  else if(p > pStart+3 && *(p-4) == UC_SENTINEL)
-  {
-    switch(*(p-3))
-    {
-      case CODE_IFOPT:
-      case CODE_IFSYSTEMSTORE:  // I3432
-        return p-4;
-    }
+  
+  q  = p - 2;
+  for (int i = 2; i < CODE__SIZE_MAX && q >= pStart; i++, q--) {
+    //  *q == UC_SENTINEL &&  *(q + 1) is within CODE__SIZE && next CODE_ right of UC_SENTINEL ( looked up in CODE__SIZE+1) has value i
+    if (*q == UC_SENTINEL &&  *(q + 1) <= CODE_LASTCODE     && CODE__SIZE[*(q + 1)] + 1 == i)
+      return (p - i);
```

Comment:
```suggestion
    if (*q == UC_SENTINEL &&  *(q + 1) <= CODE_LASTCODE     && CODE__SIZE[*(q + 1)] + 1 == i) {
      return q;
    }
```

Reply:
**### _I am also concerned that the Windows Core build passed where the other platforms didn't. We need to find out why._**

I have pushed again, in case there were problems with my files. Also I have 2 more ideas:

Incxstr/decxstr are only changed in **\common\core\desktop\src\kmx\kmx_xstring.cpp** so far. They are not changed in **\windows\src\global\vc\xstring.cpp** yet.

There are changes in pointer movement in the new version of decxstr so some tests will not give the same results as CODE__SIZE[] for new decxstr uses different values of  pointer movement compared to the old version of decxstr. This applies to 3 cases : 

- o	0x0A ( not specified in old decxstr <-> CODE__SIZE[0x0A]= 2 in new decxstr) 
- o	0x0C (not specified in old decxstr  <-> CODE__SIZE[0x0C]= 1 in new decxstr) 
- o	0x0E (specified  as 1 in old decxstr <-> CODE__SIZE[0x0E]=0 in new decxstr) 

__________________________________________________________________________________

_**And given that the code for that UC_SENTINEL is out of range, I think the test is wrong; it should be…**_

I am puzzled by the behavior using 0xF000. Here are the results of the original decxstr-function: 


- p_start = (PKMX_WCHAR)u"abc\uFFFF\U0001F609d";	 ------> assert(q == (p - 2));
- p_start = (PKMX_WCHAR)u"abc\uFFFFzd"; ------> assert(q == (p - 2)); 
- p_start = (PKMX_WCHAR)u"abc\uFFFF\uF000d";	 ------> assert(q == (p - 1));


If we decxstr a surrogate pair in the original decxstr-function, the pointer moves 2 to the left. When using characters like u"abc\uFFFFzd"  in the original function the pointer will be moved 2 to the left. If we take 0xF0000 instead of z the pointer is supposed to move one to the left?  Is there a difference in the use of characters and characters of the private use area and how would one discern them? Are they always used after UC_SENTINEL?  Or do we simply move all characters >0xF000 just one step to the left?

To not produce errors in tests I removed the tests for 0xF000 until I am really sure how to process those characters. 
__________________________________________________________________________________

Reply:
> To not produce errors in tests I removed the tests for 0xF000 until I am really sure how to process those characters.

See my comment here: https://github.com/keymanapp/keyman/pull/5842#discussion_r749770382. With that in place, we should be able to re-enable tests.

Reply:
p_start = (PKMX_WCHAR)u"abc\uFFFF\U0001F609d"; ------> assert(q == (p - 2));
p_start = (PKMX_WCHAR)u"abc\uFFFFzd"; ------> assert(q == (p - 2));
p_start = (PKMX_WCHAR)u"abc\uFFFF\uF000d"; ------> assert(q == (p - 1));

What I don`t understand is:
-  in the second line we start out at "d", move 2 to the left and end up at "FFFF" (and skip "z" ?) 
-  in the third line we start out at "d", move 1 to the left and end up at "F000"

Why do we treat these similar cases differently? 
Why do we want to end **AT** FFFF in one case and **ONE RIGHT TO** FFF in the other?
Shouldn`t we stop  **ONE RIGHT TO** FFF in both cases? Was there a reason to set it up like that in the original code?

Reply:
You are right. The behaviour was poorly defined for invalid sequences.

I have gone ahead and defined the behaviour when we have an invalid UC_SENTINEL sequence:

1. UC_SENTINEL followed by an invalid code: treat as individual characters
2. UC_SENTINEL followed by a valid code but insufficient number of parameters: treat as individual characters

Note that scenario 2 can arise when we call `decxstr` on a pointer to any parameter to the code sequence. In a correctly formed string, this should never arise., so for example, in this scenario, we are pointing to the parameter for CODE_ANY:

```c++
  p_start = (PKMX_WCHAR)u"\uFFFF\u0001\u0002";
  p = p_start + 2; // \u0002 in the middle of the otherwise valid UC_SENTINEL CODE_ANY sequence
  q = decxstr(p, p_start);
  assert(q == p - 1);
```

However, note the outcome of the following:

```c++
  p_start = (PKMX_WCHAR)u"\uFFFF\u0001\u0001";
  p = p_start + 3; // nul, i.e. at the end of the valid UC_SENTINEL CODE_ANY sequence
  q = decxstr(p, p_start);
  assert(q == p - 3);
```

In this situation, we are starting at the end of a valid UC_SENTINEL sequence, so `decxstr` correctly jumps back to the start of the sequence.

I pushed an update that applies that fix and updates the tests, so you'll want to pull that locally and verify that it makes sense to you :grin: